### PR TITLE
Reuse blocked evaluations and handle unblock events that occurred during scheduling

### PR DESCRIPTION
**Pull request by dadgar**
This PR:
* Reuses existing blocked evaluations
* Checks if there were unblock events that would have been of use to the blocked evaluation while it was in the scheduler
* Periodically unblocks failed evaluations that hit the max Plan attempts. 

@armon for review. Should be reviewed after #1188 is merged and this rebased.

### CODE REVIEW

File: nomad/blocked_evals.go [hunk 1]
```diff
@@ -133,6 +146,14 @@ func (b *BlockedEvals) Block(eval *structs.Evaluation) {
 		return
 	}
 
+	// Check if the eval missed an unblock while it was in the scheduler at an
+	// older index.
```
Comment (armon) [inline nomad/blocked_evals.go]: I would add some more comment as to how/why this is possible. "The scheduler could have been invoked with a snapshot of state that was prior to additional capacity being added or allocations becoming terminal."

File: nomad/blocked_evals.go [hunk 2]
```diff
@@ -188,6 +258,11 @@ func (b *BlockedEvals) unblock(computedClass string) {
 		return
 	}
 
+	// Store the index in which the unblock happened. We use this on subsequent
+	// block calls in case the evaluation was in the scheduler when a trigger
+	// occured.
+	b.unblockIndexes[computedClass] = index
```
Comment (armon) [inline nomad/blocked_evals.go]: If we are doing this in `Unblock` do we need to do it again?

File: nomad/leader.go [hunk 3]
```diff
@@ -341,6 +344,21 @@ func (s *Server) reapDupBlockedEvaluations(stopCh chan struct{}) {
 	}
 }
 
+// periodicUnblockFailedEvals periodically unblocks failed, blocked evaluations.
+func (s *Server) periodicUnblockFailedEvals(stopCh chan struct{}) {
+	ticker := time.NewTimer(1 * time.Minute)
```
Comment (armon) [inline nomad/leader.go]: I would move this value to a constant at the top

File: nomad/leader.go [hunk 4]
```diff
@@ -341,6 +344,21 @@ func (s *Server) reapDupBlockedEvaluations(stopCh chan struct{}) {
...
+	for {
+		select {
+		case <-stopCh:
+			ticker.Stop()
```
Comment (armon) [inline nomad/leader.go]: Just defer the stop and do a return

**Comment by armon** [comment]
LGTM

File: nomad/blocked_evals.go (hunk 2 above)
Comment (dadgar) [inline nomad/blocked_evals.go]: We don't copy paste fail

**Pull request merged**